pkg/client/client_v2.go

```diff
@@ -317,113 +327,105 @@ func (c *clientV2) checkinRoundTrip() {
-			if err != nil {
-				if !errors.Is(err, io.EOF) {
-					c.errCh <- err
-				}
-				close(done)
-				return
-			}
-			c.sync(expected)
+		defer wg.Done()
+		defer close(readerDone)
+		expected, err := checkinClient.Recv()
+		for ; err == nil; expected, err = checkinClient.Recv() {
+			c.applyExpected(expected)
+		}
+		if !errors.Is(err, io.EOF) {
+			c.errCh <- err
 		}
 	}()
 
-	// observed state check-ins
-	checkinWrite.Add(1)
+	// observed state check-ins (writer)
+	wg.Add(1)
 	go func() {
-		defer checkinWrite.Done()
-
-		if err := c.sendObserved(checkinClient); err != nil {
-			if !errors.Is(err, io.EOF) {
-				c.errCh <- err
-			}
-			return
-		}
-
-		t := time.NewTicker(c.minCheckTimeout)
-		defer t.Stop()
-
-		for {
-			select {
-			case <-done:
-				return
-			case <-c.kickSendObservedCh:
-				if err := c.sendObserved(checkinClient); err != nil {
-					if !errors.Is(err, io.EOF) {
-						c.errCh <- err
-					}
-					return
-				}
-				t.Reset(c.minCheckTimeout)
-			case <-t.C:
-				if err := c.sendObserved(checkinClient); err != nil {
-					if !errors.Is(err, io.EOF) {
-						c.errCh <- err
-					}
-					return
-				}
-			}
-		}
+		defer wg.Done()
+		c.checkinWriter(checkinClient, readerDone)
+		_ = checkinClient.CloseSend()
 	}()
 
-	// wait for write goroutine to quit then close the client
-	checkinWrite.Wait()
-	checkinClient.CloseSend()
+	// Wait for reader and writer to finish before returning.
+	wg.Wait()
+}
 
-	// then wait for the read to quit
-	checkinRead.Wait()
+func (c *clientV2) checkinWriter(
+	checkinClient proto.ElasticAgent_CheckinV2Client,
+	done chan struct{},
+) {
+	t := time.NewTicker(c.minCheckTimeout)
+	defer t.Stop()
+
+	// Keep sending until the call returns an error
+	for c.sendObserved(checkinClient) == nil {
+
+		// Wait until the ticker goes off, or we're notified of an update.
+		select {
+		case <-c.stateChangeObservedCh:
+			// We're sending an off-cycle update, so reset the ticker timeout
+			t.Reset(c.minCheckTimeout)
+		case <-t.C:
+		case <-done:
+			return
+		}
+	}
 }
 
 // sendObserved sends the observed state of all the units.
+// If the send produces any error except io.EOF, sendObserved
+// forwards it to the client's error channel before returning.
 func (c *clientV2) sendObserved(client proto.ElasticAgent_CheckinV2Client) error {
 	c.unitsMu.RLock()
 	observed := make([]*proto.UnitObserved, 0, len(c.units))
 	for _, unit := range c.units {
 		observed = append(observed, unit.toObserved())
 	}
+	// Check featuresIdx within the same mutex block since we
+	// need observed and featuresIdx to be synchronized with each other.
+	featuresIdx := c.featuresIdx
 	c.unitsMu.RUnlock()
 
-	c.featuresMu.RLock()
-	defer c.featuresMu.RUnlock()
-
 	msg := &proto.CheckinObserved{
 		Token:       c.token,
 		Units:       observed,
-		FeaturesIdx: c.featuresIdx,
+		FeaturesIdx: featuresIdx,
 		VersionInfo: nil,
 	}
-	if !c.versionInfoSent {
-		c.versionInfoSent = true
+	c.sendVersionInfoOnce.Do(func() {
```

Comment:
This only guarantees that we have filled the `VersionInfo` struct once, not that it has been successfully sent. It would be more robust to stop doing this once the `Send` call below succeeds when `VersionInfo` is set.

It looks like this was a problem before as well.

Reply:
Ah, good catch. And correct me if I'm wrong but it looks like it's important that we send it at least once but not very important that we never send it more than once? So I've reverted to a non-synchronized variable but moved its update to after the Send.

Reply:
Yes it needs to get through at least once, afterwards not sending it as an optimization to avoid pointless copying bytes that will never change between processes.

Here's the agent side of this where it pulls the values out:

https://github.com/elastic/elastic-agent/blob/78774a78a4a758bfd1c7a2a6f95fbed66c393d92/pkg/component/runtime/state.go#L322-L335

```go
	if checkin.VersionInfo != nil {
		if checkin.VersionInfo.Name != "" && s.VersionInfo.Name != checkin.VersionInfo.Name {
			s.VersionInfo.Name = checkin.VersionInfo.Name
			changed = true
		}
		if checkin.VersionInfo.Version != "" && s.VersionInfo.Version != checkin.VersionInfo.Version {
			s.VersionInfo.Version = checkin.VersionInfo.Version
			changed = true
		}
		if checkin.VersionInfo.Meta != nil && diffMeta(s.VersionInfo.Meta, checkin.VersionInfo.Meta) {
			s.VersionInfo.Meta = checkin.VersionInfo.Meta
			changed = true
		}
	}
```